src/leader.c

```diff
@@ -272,405 +158,382 @@ static void leaderMaybeCheckpointLegacy(struct leader *l)
+			} else {
+				int rc = sqlite3_prepare_v2(leader->conn, 
+					req->sql, -1, &req->stmt, &req->tail);
+				if (rc != 0) {
+					sm_move(&req->sm, EXEC_DONE);
+					req->status = RAFT_ERROR;
+				} else if (req->stmt == NULL) {
+					sm_move(&req->sm, EXEC_DONE);
+				} else {
+					sm_move(&req->sm, EXEC_PREPARED);
+					POST(req->stmt != NULL);
+				}
+			}
+			break;
+		case EXEC_PREPARED:
+			PRE(req->status == 0);
+			PRE(req->stmt != NULL);
+			if (req->work_cb == NULL) {
+				/* no work callback, we are done */
+				sm_move(&req->sm, EXEC_DONE);
+			} else if (sqlite3_stmt_readonly(req->stmt)) {
+				/* database in in WAL mode, readers can always proceed */
+				sm_move(&req->sm, EXEC_WAITING);
+			} else if (db->active_leader == NULL || db->active_leader == leader) {
+				sm_move(&req->sm, EXEC_WAITING);
+				db->active_leader = leader;
+			} else {
+				/* Supend as another leader is keeping the database busy,
+				 * but also start a timer as this statement should not sit
+				 * in the queue for too long. In the case the timer expires
+				 * the statement will just fail with RAFT_BUSY. */
+				req->status = raft_timer_start(leader->raft, &req->timer,
+					db->config->busy_timeout, 0, exec_timer_cb);
+				if (req->status != RAFT_OK) {
+					sm_move(&req->sm, EXEC_DONE);
+				} else {
+					sm_move(&req->sm, EXEC_WAITING);
+					queue_insert_tail(&db->pending_queue, &req->queue);
+					return exec_suspend(req);
+				}
+			}
+			break;
+		case EXEC_WAITING:
+			raft_timer_stop(leader->raft, &req->timer);
+			queue_remove(&req->queue);
+			if (req->status != 0) {
+				sm_move(&req->sm, EXEC_DONE);
+			} else {
+				sm_move(&req->sm, EXEC_RUN_BARRIER);
+				if (exec_needs_barrier(leader)) {
+					req->status = raft_barrier(leader->raft, &req->barrier, exec_barrier_cb);
+					if (req->status == 0) {
+						return exec_suspend(req);
+					}
+				}
+			}
+			break;
+		case EXEC_RUN_BARRIER:
+			if (req->status != 0) {
+				sm_move(&req->sm, EXEC_DONE);
+			} else {
+				sm_move(&req->sm, EXEC_RUNNING);
+				TAIL return req->work_cb(req);
+			}
+			break;
+		case EXEC_RUNNING:
+			sm_move(&req->sm, EXEC_DONE);
+			if (req->status == RAFT_OK) {
+				dqlite_vfs_frame *frames;
+				unsigned int nframes;
+				/* 
+				 * FIXME: If this was a xFileControl:
+				 *  - it would be callable through sqlite3_file_control
+				 *  - it would set the error for the connection (so, no translation needed here)
+				 *  - it would not be necessary to keep a vfs pointer in the db
+				 *  - it would not necessary to lookup the database by path every time.
+				 */
+				int rc = VfsPoll(db->vfs, db->path, &frames, &nframes);
+				if (rc == SQLITE_OK && nframes > 0) {
+					req->status = exec_apply(req, frames, nframes);
+					for (unsigned i = 0; i < nframes; i++) {
+						sqlite3_free(frames[i].data);
+					}
+					sqlite3_free(frames);
+					if (req->status == 0) {
+						return exec_suspend(req);
+					}
+				} else if (rc != SQLITE_OK) {
```

Comment:
Why not just `} else {` or `... + } else { impossible!();}`

Reply:
Maybe I'm not entire understanding this suggestion: the `else` branch is not impossible, but is success. Am I missing something here?

Reply:
Looks like now I've good the logic.